include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -781,6 +781,13 @@ struct SessionImpl : ConstSessionImpl<T> {
 
   void Run(const RunOptions& run_options, const IoBinding&);  ///< Wraps OrtApi::RunWithBinding
 
+  /** \brief Run the model in a separate thread.
+   * Callback will be invoked on run completion, with output values as arguments,
+   * on error, status could be used to see detail.
+   */
+  void RunAsync(const RunOptions& run_options, const char* const* input_names, const Value* input_values, size_t input_count,
+                const char* const* output_names, size_t output_count, RunAsyncCallbackFn callback);
```

Comment:
For C++ API, a `std::function` type without the `userdata` is preferred. callers have multiple ways to capture their user data.

Reply:
How do we pass a std::function through C API?

Reply:
we cannot. in C API we need the traditional function pointer type.

Reply:
Oh, I see what do you mean.

perhaps need to figure out how to "new" a `std::function` object and pass it as `userdata` to the callback function implementation, convert the `userdata`(void*) into `std::function` object (reference or pointer), call it and delete it.

Reply:
I would rather that we use a function ptr for C++ API as well. It avoids any unnecessary complications with std::function.

Reply:
Updated accordingly - by some bridging we could do it.